### ISSUE TITLE
Update style guide for <label> best practices

### ISSUE THREAD

**Issue by CITguy** [rackerlabs/encore-ui#948]
It is considered best practice to have a `<label for="inputId">` associated with checkbox and radio inputs (by clicking the label, the input is activated).  This should be documented in our style guide.

**Checkbox w/ Label**
```html
<input type="checkbox" id="chkDemo" />
<label for="chkDemo">Demo Checkbox</label>
```

**Radio w/ Label**
```html
<input type="radio" id="radDemo" />
<label for="radDemo">Demo Radio</label>
```

**Comment by megalord** [comment]
It's also valid to have a single `input` inside a `label`, which eliminates the need for the `for` and `id` attributes.
```html
<label>
  <input type="checkbox" />
  Demo Checkbox
</label>
```

**Comment by CITguy** [comment]
Interesting, I wasn't aware of that trick. :+1:

**Comment by CITguy** [comment]
So, while the second style is easier to implement, it brings challenges when attempting to style the label text based on whether the input is disabled.

**With the "label for" method**
```html
<style>
  /* Sibling selector allows us to style the label */
  input[disabled] + label { color: #ccc; }
</style>
<input type="checkbox" id="chkDemo" disabled />
<label for="chkDemo">Disabled Checkbox</label>
```

**Using the "label around" method"**
```html
<label>
  <input type="checkbox" disabled />
  How do I style this text when the input is disabled?
</label>
```
CSS Selectors only allow to style sideways or downward, never upward (which is needed in this case). This markup makes it impossible to style the label text.

**Comment by CITguy** [comment]
In order to perform CSS sibling selector, you'd require more markup.

```html
<style>
  /* Sibling within label */
  label input[disabled] + label span { color: #ccc; }
</style>
<label>
  <input type="checkbox" disabled />
  <span>I can style this, but it requires more markup</span>
</label>
```

**Comment by CITguy** [comment]
Two goals are desired:

1. Clicking the label should activate the control
2. The label should be styleable based on if the input is disabled.

```LESS
input[type="checkbox"] {
	&[disabled] {
		cursor: not-allowed;

		/* OPTION 1 - Sibling Label
			<input type="checkbox" id="chkDemo" disabled />
			<label for="chkDemo">Checkbox 1</label>
		
		  PROS:
		    * Specific CSS Styles - Low chance of 
		      unwanted sibling <label> styling
		  CONS:
		    * Additional markup required
		*/
		& + label {
			color: red;
			cursor: not-allowed;
		}

		/* OPTION 2 - Sibling Span
			<label>
				<input type="checkbox" disabled />
				<span>Checkbox 2</span>
			</label>
		
		  PROS:
		    * No need for `for="chkId"`
		  CONS:
		    * Too Generic CSS - High chance of 
		      unwanted sibling <span> styling
		    * Additional markup required
		*/
		& + span {
			color: blue;
			cursor: not-allowed;
		}

		/* OPTION 3 - Plain text sibling
		  <label>
		    <input type="checkbox" disabled />
		    Checkbox 3
		  </label>
		
		  PROS:
		    * Simple Markup
		  CONS:
		    * No way to style "Checkbox 3" with existing
		      CSS Selectors - requires additional markup
		*/
	}
}
```

OPTION | Click to Enable? | Style if Disabled?
------ | ------ | ------
Sibling Label | Yes | Yes
Sibling Span | Yes | Yes
Plain text sibling | Yes | No

Based on inability to accomplish the two goals, "OPTION 3 - Plain text sibling" is not desirable. Which brings us to the other two options.

Excluding that both remaining options require additional markup, "OPTION 1 - Sibling Label" is more desirable because it has less of a chance of CSS accidentally styling sibling elements when the input is disabled. Because `<label>` is less likely to appear in markup than `<span>` it is more specific and is more desirable as a CSS selector.

Additionally, considering that QE uses the `id` attribute to properly interact with form controls via automated tests, the only remaining markup would be the `for="chkId"` on the label.

With this information, I would highly recommend making "OPTION 1 - Sibling Label" the best practice within the framework.

**Comment by CITguy** [comment]
@glynnis, @ElementE, @parlarjb : any input?

**Comment by CITguy** [comment]
@Droogans too (in regards to QE remark and selecting elements for interaction)

**Comment by CITguy** [comment]
FYI: `rxCheckbox` and `rxRadio` may need their LESS files updated to reflect the decision made here. Initially, the two directives will support "OPTION 1 - Sibling Label".

**Comment by Droogans** [comment]
@CITguy although it's not pretty, this is used in a couple places in the component page objects.

```js
eleSibling: {
    get: function () {
        return this.rootElement.element(by.xpath('..')).$('.sibling');
    }
}
```

I don't think it looks good, but it's properly abstracted and nearly invisible to end users, so it ranks pretty low on my list of grievances right now.

**Comment by CITguy** [comment]
@Droogans: Internally, within the directive template, using a `.sibling` class (or similar) should be fine.  The discussion here is about best practices for consumers (outside of directive templates). Based on your QE work outside of the framework, do you have any comments or concerns about OPTION 1 or OPTION 2?

**Comment by Droogans** [comment]
Just a note -- I try to never use `id` unless there's no binding, model, or some CSS selector that reads much easier than `element(by.id('icouldbeanywherelol'))`.

**Comment by parlarjb** [comment]
I like OPTION 1 as well. What do you think about adding additional support to `rxCheckbox` and `rxRadio` to do this automatically? i.e. provide the label text as an attribute to them, and have the directives create the `<label>` elements?

**Comment by CITguy** [comment]
I'd rather not. My biggest concern against that is support for custom markup in the label.

#### Bad
```html
<input rx-checkbox 
    ng-model="foobar" 
    label="Foobar <i class='fa fa-fw fa-exclamation'></i>" />
```
This is hideous. The only way to clean this up would be to define the label markup in javascript. It is much easier on the consumer to keep as much markup in the views as possible.

#### Good
```html
<input rx-checkbox id="chkFoobar" ng-model="foobar" />
<label for="chkFoobar">
  Foobar <i class="fa fa-fw fa-exclamation"></i>
</label>
```
This is a cleaner implementation and provides more flexibility for developers. In addition, markup doesn't need to be defined in javascript for this to work.

**Comment by parlarjb** [comment]
Good point, I wasn't thinking about custom markup. :+1:

**Comment by CITguy** [comment]
Yeah, another use case would be help tooltips suggested in #932.

**Comment by CITguy** [comment]
Closing. Addressed with rxForm 2.0 updates (http://rackerlabs.github.io/encore-ui/#/styleguide/forms)
